feat(tui): colorized underline in Windows Terminal

ajmandourah opened the pull request:
Closes https://github.com/neovim/neovim/issues/26744

Windows did not support underline syles other than the regular one. no undercurl or double line . but with the latest merge these has been implemented and planned for https://github.com/microsoft/terminal/pull/16097.

in the latest Preview you can see this in action using powershell >= 6 (Cmd as far as i know does not support escape chars. correct me if i am wrong )

using both
```
Write-Host "`e[4:3mtest`e[0m"
Write-Host "`e[4:3;58:2::000:255:000mtest`e[0m"
```
will print both colored non colored undercurl

![292871937-1c559811-e124-4354-8d67-cacc0f3f78a2](https://github.com/neovim/neovim/assets/27051374/71d3c54f-ed0d-4ad5-91a3-9c12e61bd2b9)

Adding the paramaters in the terminfo for vtpcon will give full support for underline style rendering in native windows. I assume this will also give support to other terminals other than windows terminal like wezterm and alacritty as  these are relient on windows terminal conptty implementation which should update as soon as the preview branch merge into the stable one. 

To test this out download the latest Windows terminal Preview from either github or Windows store.

![image](https://github.com/neovim/neovim/assets/27051374/c84e089e-2d0b-480a-a012-1cb6de8b0871)

Thanks to @tusharsnx and @kmoschcau for the inspiration.

ajmandourah committed:
feat(tui): Undercurl support + color on windows

ajmandourah committed:
Undercurl support + color on windows

ajmandourah committed:
feat(tui): undercurl + color

ajmandourah committed:
undercurl + color

ajmandourah committed:
undercurl and color support for native windows

ajmandourah committed:
feat(tui): undercurl and color support for native windows

zeertzjq commented on src/nvim/tui/terminfo_defs.h:
```diff
@@ -1,8 +1,14 @@
 
-#pragma once
+#ifndef NVIM_TUI_TERMINFO_DEFS_H
+#define NVIM_TUI_TERMINFO_DEFS_H
```
?

ajmandourah replied:
Sorry seems like I mixed up the master br with the stable one as I could not compile the master one correctly with msvc for some reason hence I revert to stable.

Will fix this

ajmandourah committed:
fix: pragma once, generated with updated ncurses

ajmandourah committed:
fix: pragma once, generated with updated ncurses

zeertzjq commented on src/nvim/tui/terminfo_defs.h:
```diff
@@ -1,6 +1,6 @@
 // uncrustify:off
 
-// Generated by scripts/update_terminfo.sh and ncurses 6.4.20230520
+// Generated by scripts/update_terminfo.sh and ncurses 6.3.20211021
```
Why is it still 6.3 here if you generated with updated ncurses? Although I ran `./scripts/update_terminfo.sh` on your branch and the only change is the version here.

ajmandourah replied:
Honestly i don't know. I am using wsl to run the script, maybe something is wrong there. i will regenerate again with cygwin

zeertzjq replied:
Hmm that seems to cause more problems. Let's revert the newest commit then.

ajmandourah committed:
fix: wrong ncurses version

ajmandourah committed:
fix: wrong ncurses version.

zeertzjq committed:
Revert "fix: wrong ncurses version."

This reverts commit 6f250095e1849410dabef5d46b0d4c64fe40379f.

zeertzjq committed:
Revert "fix: wrong ncurses version."

This reverts commit 6f250095e1849410dabef5d46b0d4c64fe40379f.

zeertzjq commented on scripts/windows.ti:
```diff
@@ -59,6 +59,8 @@ vtpcon|ANIS emulation for console virtual terminal sequence with libuv,
 	sgr=\E[0%?%p1%p3%|%t;7%;%?%p2%t;4%;%?%p6%t;1%;m,
 	sgr0=\E[0m, smam@, smglp@, smgrp@, smglr@, smir@, smkx@, smm@, tbc@, u6@, u7@, u8@, u9@,
 	Cr@, Cs@, Ms@, XM@, kDC3@, kDC4@, kDC5@, kDC6@, kDC7@,
+        Smulx=\E[4:%p1%dm,
+        Setulc=\E[58::2::%p1%{65536}%/%d::%p1%{256}%/%{255}%&%d::%p1%{255}%&%d%;m,
```
```suggestion
	Smulx=\E[4:%p1%dm,
	Setulc=\E[58::2::%p1%{65536}%/%d::%p1%{256}%/%{255}%&%d::%p1%{255}%&%d%;m,
```

zeertzjq committed:
fixup: indent

zeertzjq committed:
fixup: indent

zeertzjq committed:
fixup: keep ncurses version unchanged

zeertzjq committed:
fixup: keep ncurses version unchanged

ajmandourah commented:
Thank you BTW for your time and for reviewing. And sorry for the mess created earlier on.

tusharsnx commented:
@ajmandourah Thanks for doing this! 


An important point we need to discuss is

Have you tested this on terminals that do not support colored underlines (`SGR 58/59`) and extended underline styles (`SGR 4:X`)?

The problem with these sub parameter (colon `:` syntax) based control sequences are, if we send the sequence to unsupported terminals, they'll not render anything.

Neovim, when not using `SGR 4:X` and `SGR 58`, sends standard `SGR 4` for all underlines (except double underline `SGR 21`), and don't even send `SGR 58` at all. This is important to keep unsupported terminals working as usual.

Unfortunately, feature detection in terminals isn't the best. You could've used [DECRQSS](https://vt100.net/docs/vt510-rm/DECRQSS.html) and [DECRPSS](https://vt100.net/docs/vt510-rm/DECRPSS.html), but that'll let you know the capability of backing conhost and not of the terminal user is interacting with 🥲

So, I think we should add a config like `compatibility.legacy_underline_styles` to help user revert to older behavior (i.e only `SGR 4` and no `SGR 4:X`). Sending `SGR 58` to unsupported terminals won't hurt I guess. I'm also not an nvim expert, so maybe there's already a way to turn off extended styles. In that case, mentioning that in the docs should be enough 🙂

ajmandourah commented:
@tusharsnx 

You are right.  here is a comparision

left: with current modification. current windows terminal (underline styles + colors not supported yet)
middle: with old terminfo. current windows terminal (underline styles + colors not supported yet)
right: with modifications. windows terminal preview (with support for underline styles and colors)

![image](https://github.com/neovim/neovim/assets/27051374/8f8e815e-aa3d-4c0b-b3d0-02ab2266876e)

from my understanding other terminals like Putty for example have their own terminfo defined in the terminfo_defs.h I made sure not to touch those .

tusharsnx commented:
So, what we are looking for is to get _left_ to work as _middle_ with the current changeset. Can we edit `Smulx` (or something similar) to get left to still show all underlines?

ajmandourah commented:
> So, what we are looking for is to get _left_ to work as _middle_ with the current changeset. Can we (as a user) edit `Smulx` (or something similar) to get the terminal on the _left_ to show the missing underlines?

You are probably right but the question is if we really need to do so . I have just tried alacritty and wezterm, 2 of the most popular terminals for windows and they both shows the correct underline (middle). As Windows terminal will eventually will support underlines when they merge it into the stable which i recon to be soon, I doubt we will need to implement a feature which may not be used. 

![image](https://github.com/neovim/neovim/assets/27051374/5d1d75b4-20a4-4ac3-b6ad-5df91de96de8)

again . other terminals like putty and conemu have their own terminfo defined in the header file so these probably wont be affected. did not try though .

tusharsnx commented:
> You are probably right but the question is if we really need to do so.

I think we need it because it's considered as a breaking change for unsupported terminals. This is probably why Neovim blocks this feature unless a terminal explicitly says (through terminfo, or similar methods) that it supports these extended control sequences.

> I doubt we will need to implement a feature which may not be used

This will used as a last resort for unsupported terminals. Though, we might not need it right now 🤔

*I'm okay with what core maintainers think is right for Neovim to do here.*

zeertzjq commented:
Enabling this unconditionally doesn't seem ideal. Is it possible for a program running in Windows Terminal to get the version of the terminal?

ajmandourah commented:
> Enabling this unconditionally doesn't seem ideal. Is it possible for a program running in Windows Terminal to get the version of the terminal?

There is `wt - v`. Will come back home and test it out.

ajmandourah commented:
After reading around in the Windows terminal (WT) repo, it seems like when running WT will set the env variable `%WT_SESSION%`. neovim can probably look it up and set the terminfo fallback to it.

ajmandourah committed:
Merge branch 'neovim:master' into master

ajmandourah commented:
Added another builtin fallback win-term terminfo which is mainly for WT.
I modified the terminfo.c with the following : 
```c++
} else if (terminfo_is_term_family(term, "vtpcon") && getenv("WT_SESSION")) {
  *termname = xstrdup("builtin_win_term");
  return unibi_from_mem((const char *)windows_terminal_terminfo,
                        sizeof windows_terminal_terminfo);
```
which specificly targeting WT. with this no more conflict is there with unsupported terminals for example the legacy command line .

![image](https://github.com/neovim/neovim/assets/27051374/901210fb-2731-4009-8a4f-b301aa3e3735)

Left: with the mentioned addition of a win-term terminfo fallback
Right: without (the vtpcon only modified terminfo)

notice that stylized underlines are not supported in the command prompt terminal .

ajmandourah committed:
feat(terminfo): Added win-term terminfo for colorized underline style in windows terminal

ajmandourah committed:
feat(terminfo): Added win-term terminfo for colorized underline style in windows terminal

gpanders commented on src/nvim/tui/terminfo_defs.h:
```diff
@@ -2118,6 +2118,210 @@ static const int8_t win32con_terminfo[] = {
   26,1,52,0,15,0,15,0,125,1,106,2,119,105,110,51,50,99,111,110,124,65,78,83,73,32,101,109,117,108,97,116,105,111,110,32,102,111,114,32,108,105,98,117,118,32,111,110,32,108,101,103,97,99,121,32,99,111,110,115,111,108,101,0,0,1,0,0,0,0,0,0,0,0,0,0,0,1,1,0,-1,-1,8,0,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,8,0,64,0,-1,-1,0,0,2,0,-1,-1,-1,-1,4,0,11,0,15,0,19,0,-1,-1,30,0,47,0 ...
 };
 
+// win-term|ANIS emulation for console virtual terminal sequence with libuv,
```
```suggestion
// win-term|ANSI emulation for console virtual terminal sequence with libuv,
```

gpanders commented on scripts/windows.ti:
```diff
@@ -48,6 +48,29 @@ conemu|ANIS X3.64 and Xterm 256 colors for ConEmu with libuv,
+	kUP3@, kUP4@, kUP5@, kUP6@, kUP7@, rmxx@, smxx@, xm@,
+	use=libuv+basekey, use=libuv+exkey, use=xterm+256color, use=xterm+sl, use=xterm-new,
+
+win-term|ANIS emulation for console virtual terminal sequence with libuv,
```
```suggestion
win-term|ANSI emulation for console virtual terminal sequence with libuv,
```

ajmandourah replied:
fixed

zeertzjq commented:
https://github.com/microsoft/terminal/issues/8303 mentions `ms-terminal` as the terminfo name.

ajmandourah commented:
> [microsoft/terminal#8303](https://github.com/microsoft/terminal/issues/8303) mentions `ms-terminal` as the terminfo name.

will change the name to ms-terminal.

on a side note. neovim still uses  vtpcon even for windows terminal . so technically we are virtualizing the terminal.

Kayzels commented:
I'm very new to Neovim, so unable to provide feedback on your commits.

However, I just tested this out on my machine. I can confirm it works perfectly in both Powershell 7 and in cmd. Colored and styled underlines show when AtlasEngine is enabled, and underlines without styles show when it is not.

![image](https://github.com/neovim/neovim/assets/72696811/fd1557b6-465d-4b0b-8d82-b7660654eb8a)

ajmandourah committed:
feat(terminfo): fixed typo.

ajmandourah committed:
feat(terminfo): fixed typo

zeertzjq commented:
Another approach is to add a condition here instead of adding a terminfo entry:
```c
  // Extended underline.
  // terminfo will have Smulx for this (but no support for colors yet).
  tui->unibi_ext.set_underline_style = unibi_find_ext_str(ut, "Smulx");
  if (tui->unibi_ext.set_underline_style == -1) {
    int ext_bool_Su = unibi_find_ext_bool(ut, "Su");  // used by kitty
    if (vte_version >= 5102 || konsolev >= 221170
        || (ext_bool_Su != -1
            && unibi_get_ext_bool(ut, (size_t)ext_bool_Su))) {
      tui->unibi_ext.set_underline_style = (int)unibi_add_ext_str(ut, "ext.set_underline_style",
                                                                  "\x1b[4:%p1%dm");
    }
  }
```

ajmandourah commented:
> Another approach is to add a condition here instead of adding a terminfo entry:
> ```c
>   // Extended underline.
>   // terminfo will have Smulx for this (but no support for colors yet).
>   tui->unibi_ext.set_underline_style = unibi_find_ext_str(ut, "Smulx");
>   if (tui->unibi_ext.set_underline_style == -1) {
>     int ext_bool_Su = unibi_find_ext_bool(ut, "Su");  // used by kitty
>     if (vte_version >= 5102 || konsolev >= 221170
>         || (ext_bool_Su != -1
>             && unibi_get_ext_bool(ut, (size_t)ext_bool_Su))) {
>       tui->unibi_ext.set_underline_style = (int)unibi_add_ext_str(ut, "ext.set_underline_style",
>                                                                   "\x1b[4:%p1%dm");
>     }
>   }
> ```
> Since there are already several conditions, I'm not sure which solution is better.

Didn't go that deep to be honest. But it's as you said another possibility. Which can be added as failsafe.

I guess what really going to decide is the performance (which I think won't matter here as it is literally just a condition) and how extendable the approach in case new features are added. I don't know if there are other terminal features that WT does not support yet in comparison with Unix terminals.

I am open and up for suggestings.

ajmandourah committed:
feat(terminfo): revert Setulc

ajmandourah committed:
feat(terminfo): revert Setulc

ajmandourah commented:
after some testing it seems like adding Setulc caused some issue when scrolling around . was noticed when using nvim-tree plugin but I recon this will also affect multiple windows. Removing the parameter reserve the ability of windows terminal to render the underlines and the colors so only smlux was needed.

ajmandourah committed:
Merge branch 'neovim:master' into master

ajmandourah commented:
Is there anything left to be done here?

clason commented:
Fix the failing tests and squash your commits?

ajmandourah commented:
> Fix the failing tests and squash your commits?

thanks. I am new to git. will look it up

ajmandourah committed:
feat(tui): Merge branch 'test'

ajmandourah committed:
feat(tui): change getenv to os_getenv

ajmandourah committed:
Revert "feat(tui): change getenv to os_getenv"

This reverts commit eb7af34e08b85a68defbac6d9ab15574414fc3c5.

ajmandourah committed:
feat(tui): change getenv to os_getenv

ajmandourah commented:
I tried changing getenv() to os_getenv() as recommended by lint. it compiles fine in visual studio but does not pass the test. any ideas to proceed?

tusharsnx commented on src/nvim/tui/terminfo.c:
```diff
@@ -120,6 +122,10 @@ static unibi_term *terminfo_builtin(const char *term, char **termname)
     *termname = xstrdup("builtin_conemu");
     return unibi_from_mem((const char *)conemu_terminfo,
                           sizeof conemu_terminfo);
+  } else if (terminfo_is_term_family(term, "vtpcon") && os_getenv("WT_SESSION")) {
```
nit:
```suggestion
  } else if (terminfo_is_term_family(term, "vtpcon") && os_env_exists("WT_SESSION")) {
```

tusharsnx commented on src/nvim/tui/terminfo.c:
```diff
@@ -14,11 +14,13 @@
 #endif
 
 #ifdef INCLUDE_GENERATED_DECLARATIONS
+# include "os/env.c.generated.h"
```
```suggestion

```

You should get rid of this include and instead remove the header gaurd (`#ifdef __FreeBSD__` and `#endif`) around the `nvim/os/os.h` include. That should include the required `os_getenv()` (or `os_env_exists()` as I suggested in another comment) correctly.

ajmandourah replied:
Thank you for your suggestions. will apply the changes and test it out

tusharsnx commented on src/nvim/tui/terminfo.c:
```diff
@@ -14,11 +14,13 @@
 #include "nvim/tui/terminfo.h"
 #include "nvim/tui/terminfo_defs.h"
 
+
```
```suggestion

```

Remove empty lines.

tusharsnx reviewed (commented):
Looks great.

You can apply the suggested changes and see if it helps.

ajmandourah committed:
feat(tui): using os_env_exists instead of os_getenv

ajmandourah commented:
> Looks great.
> 
> You can apply the suggested changes and see if it helps.

it compiles in msvc. getting this on lint check

```
FAIL: /home/runner/work/neovim/neovim/src/nvim/tui/terminfo.c (File size changed from 8862 to 8861)
```

dundargoc commented:
Run `make format` locally.

ajmandourah committed:
feat(tui): using os_env_exists instead of os_getenv

ajmandourah commented:
> Run `make format` locally.

Thanks.
it passes lint now.

ajmandourah commented:
Now FreeBSD is complaingin but looks like an issue with the script.

zeertzjq commented:
There seems to still be some problems:
- The new terminfo entry is not added to `scripts/update_terminfo.sh`, so it will be removed when the script is next run.
- The new terminfo entry is mostly a copy of `vtpcon`, which may lead to future mistakes (e.g. only updating one entry when both should be updated).

So I'm starting to think https://github.com/neovim/neovim/pull/27629#issuecomment-1975770891 is a better approach. I've opened a PR #28035 for that.

zeertzjq commented:
Hmm, it seems that checking `WT_SESSION` environment variable still prevents undercurl from working in older Windows Terminal versions. I checked this in Windows Terminal 1.12.10983.0, and highlights with undercurl show neither underline or undercurl.

ajmandourah commented:
> Hmm, it seems that checking `WT_SESSION` environment variable still leads to worse behavior on older Windows Terminal versions. I checked this with Windows Terminal 1.12.10983.0, and a highlight group with undercurl shows neither an underline nor an undercurl, while without `WT_SESSION` it shows an underline.

Undercurl supported since 1.19. You are right older version will have this issue as there is no way to get the version installed through the environment variables. The only way I know of is through windows registry but this will be so messy and prone to changes from Microsoft side.

Updating windows terminal is the only way to go for now.
I will be on the lookout for ways to check what version is installed but seems like the devs are not interested into providing this information through env vars.

zeertzjq commented:
Also related: https://github.com/microsoft/terminal/issues/1040
